contracts/CDPLeverage.sol

```diff
@@ -0,0 +1,456 @@
+        _;
+    }
+
+
+    // Events
+    event OpenPosition(address owner, uint256 principalEth, uint256 purchPrice, uint256 layers);
+    event ClosePosition(address owner, uint256 sellPrice);
+
+
+    /**
+    * @dev Constructor function
+    * Collects module information from the tub and matchingMarket (OasisDex)contracts
+    */
+    constructor()
+        public
+    {
+        address _tub = 0xa71937147b55Deb8a530C7229C442Fd3F31b7db2;         //Kovan
+        address _oasisDex = 0x8cf1Cab422A0b6b554077A361f8419cDf122a9F9;    //Kovan
+        tub = ITub(_tub);
+        weth = tub.gem();
+        peth = tub.skr();
+        dai = tub.sai();
+        pip = tub.pip();
+        pep = tub.pep();
+        gov = tub.gov();
+        fee = tub.fee();
+        axe = tub.axe();
+        dex = IMatchingMarket(_oasisDex);
+
+        makerLR = ray2wad(tub.mat());
+
+        // we approve tub, tap, and dex to access weth, peth, and dai contracts
+        weth.approve(tub, 1000000000000000000000000);
+        weth.approve(dex, 1000000000000000000000000);
+        peth.approve(tub, 1000000000000000000000000);
+        dai.approve(tub, 10000000000000000000000000000);
+        dai.approve(dex, 10000000000000000000000000000);
+        gov.approve(tub, 1000000000000000000000000);
+        gov.approve(dex, 1000000000000000000000000);
+    }
+
+
+    // Fallback function
+    function() payable public {}
+
+
+    /**
+    * @dev Leverage - Creates a leveraged long USD/ETH position
+    * Principal (message value) of sender is used to:
+    * 1. Open CDP
+    * 2. Lock PETH into a CDP
+    * 3. Draw DAI and trade for WETH on OasisDEX
+    * 4. Convert WETH to PETH
+    * 5. Repeat steps 2 - 4 for # of (layers-1) prescribed
+    * 6. Holds CDP ownership and remaining DAI
+    * @param _priceFloor - position liquidated if USD/ETH price drops below this point
+    * @param _layers - # of layers used to leverage
+    */
+    function leverage(uint256 _priceFloor, uint256 _layers)
+        payable
+        public
+    {
+        uint256 currPriceEth = uint256(pip.read());                                   // DAI/WETH In WAD
+        uint256 currPricePeth = wmul(currPriceEth,wdiv(weth2peth(1*WAD),(1*WAD)));    // DAI/PETH In WAD
+
+        // ensure that the price floor is less than the current price of eth
+        require(_priceFloor < currPriceEth, "Price floor should not be less than the Dai/Eth price feed.");
+        require(0 < msg.value, "Ether is required to open a position.");
+        require(investors[msg.sender].principal == 0, "Previous position must be liquidated before opening a new one");
+
+        // calculate collateralization ratio from price floor
+        uint256 collatRatio = wdiv(wmul(currPriceEth, makerLR),_priceFloor);
+
+        //find exchange rate for weth/Dai
+        uint256 wethAmt;
+        uint256 daiAmt;
+        (wethAmt, , daiAmt, ) = dex.getOffer(dex.getBestOffer(weth, dai));
```

Comment:
I think it would be more clear if we had an intermediate variable called offerID for the return value of getBestOffer

Reply:
yea, you're right. I think I remember doing this to cut down the local variable cap per function (I think it's 16 or so). Note: this logic will change to report a more accurate avg. exchange rate per open position